Handle nickserv on reconnect

dedeibel opened the pull request:
Please accept my PR, since it might be of use to others.

The following changes solve the issue of being able to reauthenticate and rejoin a restricted channel after a disconnect on https://hackint.org/services where you have to request an invite to such a channel.

0. When you have a disconnect you will eventually reconnect with a different nick until the old user is removed.
1. keepnick will notice that and change the nick back
2. nickserv will handle authentication
2.1. the new postcmd allows to ask ChanServ for an invite to the restricted channel
3. invitejoin will accept the invite and join the restricted channel
3.1 the new allowed users list will only accept invite from ChanServ ...

I tried to use the already present style of loading / storing settings and extended the help dialogs.

ailin-nemui commented:
I recommend that you use sasl or certfp instead

dedeibel commented:
Thanks, SASL it is. Nonetheless The code works and is useful if nickserv is used.

ailin-nemui commented on scripts/invitejoin.pl:
```diff
@@ -33,41 +33,213 @@
+    @allowed_nicks = ();
+    if (-e $file) {
+        local *F;
+        open(F, "<", $file);
```
while you're in the process of modifying the script, why not replace this with "modern style" `open my $f, "<", $file`

dedeibel replied:
borrowed it from the nickserv script actually ...

ailin-nemui commented on scripts/invitejoin.pl:
```diff
@@ -33,41 +33,213 @@
+
+    if (-e $file) {
+        local *F;
+        open(F, ">", $file);
```
this function is weird, why create the file and call yourself recursively instead of just saving it?

dedeibel replied:
borrowed it from the nickserv script actually ...

I could make it nicer if you say there is a chance of merging.

ailin-nemui replied:
what makes  you think it has no chance of being accepted?

dedeibel replied:
Just being unsure about how this repo is maintained and who considers what feature useful.

ailin-nemui commented on scripts/invitejoin.pl:
```diff
@@ -33,41 +33,213 @@
+sub is_allowed_nick {
+  my ($server, $nick) = @_;
+
+  return 1 unless @allowed_nicks < 1;
```
so this will always return true if there are *any* allowed nicks, but always fail if there are no allowed nicks? or how to understand the code?

dedeibel replied:
Not quite.

It keeps the previous behavior which allows all invites without restriction. If you have set any allowed nicks, it restricts the invite joins to the allowed nicks.

> so this will always return true if there are**n't** any allowed nicks, but always fail if there are no **matching** allowed nicks?

ailin-nemui replied:
I wonder if that is really the case...

ailin-nemui replied:
`unless @allowed_nicks < 1` that means `IF @allowed_nicks >= 1 ` then it should (unconditionally) return true, or am I mistaken? please double check this

dedeibel replied:
unless got me again ... when you thought you could handle it

Fixing it. My testcase must have been bad too.

ailin-nemui commented on scripts/nickserv.pl:
```diff
@@ -442,13 +555,25 @@ sub nickserv_notice {
         } elsif ($text =~ /^Password accepted - you are now recognized/ || $text =~ /^You are now identified for/) {
             Irssi::signal_stop();
             Irssi::printformat(MSGLEVEL_CLIENTCRAP, 'password_accepted', $server->{tag});
+	    run_postcmds($server, $server->{tag}, $server->{nick})
```
nitpick, the indentation seems off

ailin-nemui commented on scripts/nickserv.pl:
```diff
@@ -545,18 +693,27 @@ sub nickserv_runsub {
 Irssi::command_bind('ns delnick', 'del_nickname');
 
+Irssi::command_bind('nickserv delpostcmd', 'del_postcmd');
+Irssi::command_bind('ns delpostcmd', 'del_postcmd');
```
I can't say I like all those command_binds very much... alas they were there already

ailin-nemui replied:
maybe I would remove the binds for /ns nevertheless. any user could manually /alias them. another reservation I have about this script, is that /nickserv is often /alias msg nickserv by some people. but then again those don't have to install the script

dedeibel commented:
Thanks for looking into it. I'll come up with some improvements - need to do some tests.

dedeibel commented:
Fixed the bug you metioned, did some refactorings, added some missing checks, did some more testing.
The ns command is still in there, didn't want to change previously working behavior.

ailin-nemui commented:
by the way did you know that irssi has an orthogonal feature of autojoin on invite to auto channels?

dedeibel commented:
oh ... well, thanks. Must have overlooked it.

ailin-nemui commented:
(it looks like it has been broken to start with / not caused by your changes at least, so I wonder since you changed the script, could you even use it?)

dedeibel commented:
about the pager.pl ... I did not really use it. Just looked into it, saw it's formatting was totally mad, ran vim autoindent and left it there. Maybe I should have left it untouched I guess.

ailin-nemui commented:
yea.. in that case please remove it from your commit

dedeibel commented:
it's gone, `git pull -f` in case you had a checkout of the branch

dedeibel commented:
Wohoo, thx
